### DISCUSSION TITLE
What are "bytes allocated in object heap" as reported by process_stats?

### DISCUSSION THREAD

**Discussion by kasperl** [toitlang/toit#476] [Q&A]
According https://libs.toit.io/core/utils/library-summary#process_stats process_stats shell return bytes allocated in object heap.
It seems, my program allocate more bytes than the ESP32 has RAM available. Why does the bytes allocated in object heap increase continuously?

```
jag version
Version:         v1.0.7
SDK version:     v1.6.7
Build date:      2022-02-18T14:26:00Z
```
```
main:
  i := 0
  while true:
    if i % 100 == 0:
      print "Test: $i GC#: $gc_count Heap: $process_stats[4]"
    i++
    sleep --ms=1
```
```
Test: 0 GC#: 0 Heap: 1788
Test: 100 GC#: 0 Heap: 4832
Test: 200 GC#: 0 Heap: 4980
...
Test: 9500 GC#: 0 Heap: 19424
Test: 9600 GC#: 0 Heap: 19580
Test: 9700 GC#: 0 Heap: 19736
Test: 9800 GC#: 0 Heap: 19892
Test: 9900 GC#: 0 Heap: 20048
Test: 10000 GC#: 1 Heap: 20204
Test: 10100 GC#: 1 Heap: 20360
Test: 10200 GC#: 1 Heap: 20516
Test: 10300 GC#: 1 Heap: 20672
Test: 10400 GC#: 1 Heap: 20828
Test: 10500 GC#: 1 Heap: 20984
...
Test: 21400 GC#: 1 Heap: 37988
Test: 21500 GC#: 1 Heap: 38144
Test: 21600 GC#: 1 Heap: 38300
Test: 21700 GC#: 2 Heap: 38456
Test: 21800 GC#: 2 Heap: 38612
Test: 21900 GC#: 2 Heap: 38768
...
Test: 684800 GC#: 59 Heap: 1096284
Test: 684900 GC#: 59 Heap: 1096444
Test: 685000 GC#: 60 Heap: 1096604
Test: 685100 GC#: 60 Heap: 1096764
...
```

(based on https://github.com/toitlang/toit/issues/473)

**Comment by kasperl** [comment]
The "bytes allocated in the heap" tracks the total number of allocations, but doesn't deduct the sizes of objects that die. So it is a way to follow the allocation pressure of your process. You can also use [bytes_allocated_delta](https://libs.toit.io/core/utils/library-summary#bytes_allocated_delta(0%2C0%2C0%2C)):

```
main:
  bytes_allocated_delta  // Reset counter.
  ...
  bytes := bytes_allocated_delta
  print "Operation allocated $bytes bytes"
```

If you want to know the heap size of your process, use entries 1 and 2 of the `process_stats` array. Entry 1 is the combined size of the live objects in your heap, and entry 2 is the size of the heap.

**Comment by muhlpachr** [comment]
Thank you. It would be nice to add that info into process_stats documentation.
How can I monitor heap for whole VM (reserved space, allocated vs free) ?
Something like serial_print_heap_report but with return values accessible form program.
How can I get memory allocated by process off-heap e.g. by create_off_heap_byte_array ?
